### DISCUSSION TITLE
help w/ partial autocomplete

### DISCUSSION THREAD

**Discussion by rramsey** [chrisant996/clink#477] [Q&A] [answered]
Hi,

Been using clink for a year or so and really loving it.  And now I have a question about auto-completing commands.  I often use very similar long commands and I'd like to be able to use the arrow keys to autocomplete a command letter by letter.  I'm on windows using Terminal Preview with powershell.  Let's say I have the following command in my history:  `foo bar baz dir=c:\foobar\bar --some-switch --some-other switch`.

I now need to run `foo bar baz dir=c:\bazfoo\foo --some-switch --some-other switch`.  I type `fo` and clink shows me the autocomplete.  All perfect.  I hit tab or the right arrow key and it autocompletes to `foo bar baz dir=c:\foobar\bar --some-switch --some-other switch` I hit backspace and have to retype the line after c:\.  What I'd like to be able to do is to type 'fo` and when I see the light grey command from autocomplete, use the right arrow key to move letter by letter through the autocomplete text.  

Ideally I'd be able to edit the autocomplete and keep the parts I want before and after the current cursor position, so I would right arrow until `foo bar baz dir=c:\` is active, type `bazfoo\foo` and then delete `foobar\bar` and then tab or right arrow to finish out the command from the autocomplete prompt.

Does that make sense or have I overcomplicated my description?  I'm thinking this would probably be a lua script that I might have to write.  I tried alt+h and looked through the commands, but didn't see the one that would do this.  

Ideas?

bob

**Comment by chrisant996** [comment, marked as answer]
> Been using clink for a year or so and really loving it. And now I have a question about auto-completing commands. I often use very similar long commands and I'd like to be able to use the arrow keys to autocomplete a command letter by letter. 

It sounds like you're talking about the [auto-suggest](https://chrisant996.github.io/clink/clink.html#gettingstarted_autosuggest) feature, not [completions](https://chrisant996.github.io/clink/clink.html#how-completion-works).

> I'm on windows using Terminal Preview with powershell. Let's say I have the following command in my history: `foo bar baz dir=c:\foobar\bar --some-switch --some-other switch`.
> 
> I now need to run `foo bar baz dir=c:\bazfoo\foo --some-switch --some-other switch`. I type `fo` and clink shows me the autocomplete. All perfect. I hit tab or the right arrow key 

Those keys perform very different actions there:
- <kbd>Tab</kbd> completes the word at the cursor using argument completion.
- <kbd>Right</kbd> accepts the entire auto-suggestion (dark gray text) from the auto-suggest feature (which by default shows the most recent matching command from history, or if no command matches then it shows the first available match from argument completion).

> and it autocompletes to `foo bar baz dir=c:\foobar\bar --some-switch --some-other switch` I hit backspace and have to retype the line after c:. What I'd like to be able to do is to type 'fo` and when I see the light grey command from autocomplete, use the right arrow key to move letter by letter through the autocomplete text.

It may be helpful to review the documentation for [Auto-suggest](https://chrisant996.github.io/clink/clink.html#gettingstarted_autosuggest).  <kbd>Ctrl</kbd>-<kbd>Right</kbd> accepts the next word from the auto-suggestion text.  Or <kbd>Shift</kbd>-<kbd>Right</kbd> accepts the next full word, delimited only by spaces.  There isn't a command to accept a single letter at a time from the auto-suggestion text (dark gray text).

The auto-suggest feature integrates with certain commands bound to certain keys, to do additional actions in cases where the commands otherwise wouldn't have done anything (e.g. <kbd>Right</kbd> at the end of the input line).  It isn't really feasible to add a new command for accepting a single letter at a time because of the complexities of how it's integrated as extra actions into other pre-existing commands.

P.S.  The auto-suggest feature copies how the [fish](https://fishshell.com/) shell for Linux does auto-suggest.  PowerShell also now has auto-suggest, which it calls [Predictive IntelliSense](https://devblogs.microsoft.com/powershell/announcing-psreadline-2-1-with-predictive-intellisense/).

**Reply by rramsey** [reply]
Auto-suggest is what I was looking for.  Or at least close enough to what I was thinking.

And yes, was mixing up tab and right arrow in my head.

Thank you!
